runtime/lua/vim/lsp/actions.lua

```diff
@@ -0,0 +1,343 @@
+--[=[
+
+actions.lua
+
+Define common actions that can be taken on different LSP structures.
+
+Module structure is:
+
+  actions.<lsp_structure_name>.<action_name>
+
+  Where:
+    lsp_structure_name:
+      Name of the structure as defined in the LSP specification.
+
+      For example: Location, LocationLink, Hover, etc.
+
+      All structures are assumed to be (wherever possible) to be either a singular or list of those items.
+        The LSP spec often has something return either `Location` or `Location[]`.
+        All callbacks here should be able to handle either of them.
+
+
+    action_name:
+      Name of what the user can expect to happen when this function is called.
+
+
+  Example:
+    actions.Location.jump_first will jump to the singular location passed or the first location in a list of locations.
+
+
+All actions should return a function that returns a function of that has the structure of:
+  function(err, method, params, client_id, bufnr)
+
+All actions are wrapped in a metatable that allows two conveniences:
+
+  1. When calling the `action` directly, it will perform the default action.
+
+    You can write the code directly as a call (through the use of `__call` metamethod): >
+
+      Location.jump_first(_, method, result)
+<
+
+  2. Allows for storing a callback with different configuration by using the `with` key: >
+
+      local my_highlight = Location.highlight.with { higroup = 'Substitute', timeout = 100 }
+
+<
+
+
+-- Don't make this a local variable if you want to use it in a mapping
+-- it needs to be global to be accessed easily.
+--
+-- The convention is to prefix with `on` to differentiate from other Location objects
+-- from other namespaces.
+onLocation = require('vim.lsp.actions').Location
+
+-- only `jump` to definition
+vim.lsp.buf.defintion { callbacks = onLocation.jump_first }
+
+-- `jump` to definition, and then `highlight` the item
+vim.lsp.buf.definition { callbacks = { onLocation.jump_first, onLocation.highlight } }
+
+-- `jump to definition, and then perform a `highlight` action with `higroup` set to 'Substitute' and timeout set to 2s
+vim.lsp.buf.definition {
+  callbacks = {
+    onLocation.jump_first, onLocation.highlight.with { higroup = 'Substitute', timeout = 2000 }
+  }
+}
+
+
+To explain:
+Actions should be composable
+Actions should be chainable / consecutive
+
```

Comment:
TODO:
Decide on if it should be possible for something to cancel the callbacks.
Decide if the `call` method should be defined on the objects, rather than in buf request (people maybe confused about that)?

Reply:
TODO: Add ability to do something like `if isStructure(Location, object)` or similar to make it easier to define complex callbacks? Or at least give the building blocks to do so easily.

Reply:
TODO: Add the auto checking for no results / error checking